### PR TITLE
Updated format to display File sizes

### PR DESCRIPTION
**Pull request by Hetarth02**
This PR aims to close Issue #2272.

Before:

268888423-e10b8f94-6931-4436-a570-37b6b96654d3

After:

image

### CODE REVIEW

File: src/html/HTMLWriter.jl [hunk 1]
```diff
@@ -1790,6 +1795,27 @@ function write_html(ctx::HTMLContext, navnode::Documenter.NavNode, page_html::DO
     return true
 end
 
+"""
+Calculates and converts bytes to appropriate format.
+"""
+function format_units(size, unit = "bytes")
```
Comment (odow) [inline src/html/HTMLWriter.jl:1801]: Just make this function return a `String`. It doesn't get used elsewhere

File: src/html/HTMLWriter.jl [hunk 2]
```diff
@@ -1768,11 +1768,16 @@ function write_html(ctx::HTMLContext, navnode::Documenter.NavNode, page_html::DO
...
+    file_size_format_results = format_units(file_size)
+    size_threshold_format_results = format_units(ctx.settings.size_threshold)
+    size_threshold_warn_format_results = format_units(ctx.settings.size_threshold_warn)
+
     size_threshold_msg(var::Symbol) = """
     Generated HTML over $(var) limit: $(navnode.page)
-        Generated file size: $(file_size) (bytes)
-        size_threshold_warn: $(ctx.settings.size_threshold_warn) (bytes)
-        size_threshold:      $(ctx.settings.size_threshold) (bytes)
+        Generated file size: $(file_size_format_results.size) $(file_size_format_results.unit == "" ? "" : "($(file_size_format_results.unit))")
```
Comment (odow) [inline src/html/HTMLWriter.jl:1778]:
```suggestion
        Generated file size: $file_size_format_results
```

**Review by odow — COMMENTED** [review]
Needs tests. Really just a few simple ones for `format_units`

File: src/html/HTMLWriter.jl [hunk 3]
```diff
@@ -1790,6 +1795,27 @@ function write_html(ctx::HTMLContext, navnode::Documenter.NavNode, page_html::DO
...
+        size = size / 1024
+        unit = "KiB"
+
+        if size >= 1024
+            size = size / 1024
+            unit = "MiB"
+        end
+    end
+
+    return (size = round(size, digits = 2), unit = unit)
```
[the remark this answers is not in the document]
Comment (Hetarth02) [reply src/html/HTMLWriter.jl:1816]: Since this is a utility function, I would argue against doing this, as it can be useful somewhere else in the future.

**Comment by Hetarth02** [comment]
> Needs tests. Really just a few simple ones for `format_units`

@odow I really don't know how to write tests. Can anyone guide me here?

CC: @mortenpi

File: src/html/HTMLWriter.jl (hunk 1 above)
Comment (Hetarth02) [reply src/html/HTMLWriter.jl:1801]: Since this is a utility function, I would argue against doing this, as it can be useful somewhere else in the future.

Comment (mortenpi) [reply src/html/HTMLWriter.jl:1801]: I'm with @odow here: it's better to have the string formatting complexity just once in this utility function, rather than reproduced three times in `size_threshold_msg`. The latter is not very readable right now. The function is quite simple and internal, so we can always change things up if we need to reuse this somehow.

Comment (Hetarth02) [reply src/html/HTMLWriter.jl:1801]: > The function is quite simple and internal, so we can always change things up if we need to reuse this somehow.

Alright, no problem will refactor and update the PR.

**Comment by mortenpi** [comment]
@Hetarth02 The tests for `HTMLWriter` utility functions live in https://github.com/JuliaDocs/Documenter.jl/blob/master/test/htmlwriter.jl

I thought that the Julia manual would have a tutorial on testing, but it doesn't look like it does. But these resources might help get started if you haven't touched tests before:

* https://julia.quantecon.org/software_engineering/testing.html#more-on-writing-tests
* https://docs.julialang.org/en/v1/stdlib/Test/
* https://pkgdocs.julialang.org/dev/creating-packages/#Adding-tests-to-the-package

**Comment by Hetarth02** [comment]
I have added some test. Please review them and let me know if they are what you expected.

File: test/htmlwriter.jl [hunk 4]
```diff
@@ -24,6 +24,25 @@ function verify_redirect_file(redirectfile, version)
...
+        unit = "KiB"
+        if size >= 1024
+            size = size / 1024
+            unit = "MiB"
+        end
+    end
+
+    return string(round(size, digits = 2), " (", unit, ")")
+end
+
```
Comment (mortenpi) [inline test/htmlwriter.jl:45]: The tests should test the actual function you've implemented.

```suggestion
```

File: test/htmlwriter.jl [hunk 5]
```diff
@@ -296,6 +315,11 @@ end
         @test html.size_threshold_warn == 1234
     end
 
+    @testset "HTML: size_threshold_result_formats" begin
+        @test typeof(format_units(1024)) == String
+        @test typeof(format_units(typemax(Int))) == String
```
Comment (mortenpi) [inline test/htmlwriter.jl:320]: Slightly expanding, to cover more edge cases, and also making sure (1) we test the actual function, and (2) actually testing the output (which is the interesting thing to test for):

```suggestion
        @test HTMLWriter.format_units(0) == "0 (bytes)"
        @test HTMLWriter.format_units(1) == "1 (bytes)"
        @test HTMLWriter.format_units(1023) == "1023 (bytes)"
        @test HTMLWriter.format_units(1024) == "1 (KiB)"
        @test HTMLWriter.format_units(4000) == "4 (KiB)"
        @test HTMLWriter.format_units(2^20 + 123) == "@ (MiB)"
        @test HTMLWriter.format_units(typemax(Int) == "(no limit)"
```

Not 100% sure if these tests pass actually.

File: test/htmlwriter.jl [hunk 6]
```diff
@@ -296,6 +315,11 @@ end
         @test html.size_threshold_warn == 1234
     end
 
+    @testset "HTML: size_threshold_result_formats" begin
```
Comment (mortenpi) [inline test/htmlwriter.jl:318]:
```suggestion
    @testset "HTML: format_units" begin
```

File: src/html/HTMLWriter.jl [hunk 7]
```diff
@@ -1790,6 +1795,25 @@ function write_html(ctx::HTMLContext, navnode::Documenter.NavNode, page_html::DO
     return true
 end
 
+"""
+Calculates and converts bytes to appropriate format.
+"""
+function format_units(size, unit = "bytes")
+    if size == typemax(Int)
+        return "No Limit"
```
Comment (mortenpi) [inline src/html/HTMLWriter.jl:1803]:
```suggestion
        return "(no limit)"
```

File: src/html/HTMLWriter.jl [hunk 8]
```diff
@@ -1790,6 +1795,25 @@ function write_html(ctx::HTMLContext, navnode::Documenter.NavNode, page_html::DO
     return true
 end
 
+"""
+Calculates and converts bytes to appropriate format.
+"""
+function format_units(size, unit = "bytes")
```
Comment (mortenpi) [inline src/html/HTMLWriter.jl:1801]: It doesn't really make sense to pass this as an argument here.

```suggestion
function format_units(size)
    unit = "bytes"
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * `id` anchors may now start with a numeric digit. ([#744], [#2325])
 * Documenter prints a more informative warning now if there is unexpected Julia interpolation in the Markdown (e.g. from errant `$` signs). ([#2288], [#2327])
 * Documenter now warns when it encounters invalid keys in the various key-value at-blocks. ([#2306], [#2324])
+* File sizes are now expressed in more Human-readable format. ([#2272])
```
Comment (mortenpi) [inline CHANGELOG.md:13]:
```suggestion
* File sizes are now expressed in more human-readable format. ([#2272], [#2344])
```

**Review by mortenpi — COMMENTED** [review]
Partly just a bit of bikeshedding, but also fixing the tests.

**Comment by Hetarth02** [comment]
Updated as per the feedbacks.

File: test/htmlwriter.jl [hunk 10]
```diff
@@ -296,6 +296,18 @@ end
         @test html.size_threshold_warn == 1234
     end
 
+    @testset "HTML: format_units" begin
+        @test typeof(HTMLWriter.format_units(1024)) == String
+        @test typeof(HTMLWriter.format_units(typemax(Int))) == String
```
Comment (mortenpi) [inline test/htmlwriter.jl:301]:
```suggestion
```

**Comment by Hetarth02** [comment]
@mortenpi 

Do I need to reiterate on this, or is this done?

Can I help in anyway in merging this?

**Review by mortenpi — APPROVED** [review]
No, this is good to go. Just trying to get the CI green (#2351).

**Comment by thazhemadam** [comment]
Just saw this and wanted to note that [`Base.format_bytes`](https://github.com/JuliaLang/julia/blob/0c46852901c63b33f0603b0afd58ff1da687d760/base/timing.jl#L131-L140) could/should probably have been used here.
It's significantly more performant and covers more cases as well (as unnecessary as that may be), all the while lifting the (seemingly minor) burden of maintenance off your shoulders. 🙂

**Comment by Hetarth02** [comment]
> Just saw this and wanted to note that Base.format_bytes could/should probably have been used here.
It's significantly more performant and covers more cases as well (as unnecessary as that may be), all the while lifting the (seemingly minor) burden of maintenance off your shoulders. 🙂

@mortenpi Should we reopen this?

**Comment by mortenpi** [comment]
> [Base.format_bytes](https://github.com/JuliaLang/julia/blob/0c46852901c63b33f0603b0afd58ff1da687d760/base/timing.jl#L131-L140) could/should probably have been used here.

It's good to know that it exists, and could have been used, but don't think it's worth changing now that this is merged. Something to note, though, is that `Base.format_bytes` is not public API.
